### PR TITLE
pip installable comfyui, single file .exe, multi-backend / distributed comfyui, pip installable plugins, rich API, OpenAPI specs

### PR DESCRIPTION
**Pull request by doctorpangloss**
## A fully featured update to ComfyUI

This fork allows you to

 - install ComfyUI using `pip install git+https://github.com/hiddenswitch/ComfyUI.git`. You never need to think about installing torch again.
 - build a ComfyUI .exe using `python -m PyInstaller --onefile --noupx -n ComfyUI --add-data="comfy/;comfy/" --paths . --paths comfy/cmd main.py`
 - start comfyui using `comfyui` with lots of useful options you can find in `--help`
 - run a distributed comfyui, separating the frontend from the backend, with support for unlimited backends. Start RabbitMQ, then `comfyui-worker --distributed-queue-connection-uri=amqp://guest:guest@rabbitmq-server  --cwd //shared/network/path` on as many machines as you'd like; `comfyui --distributed-queue-roles=frontend --distributed-queue-connection-uri=amqp://guest:guest@rabbitmq-server --cwd //shared/network/path` on others. 
 - author plugins with a `setup.py`, and install new nodes or web extensions with `pip install git+https://github.com/some/repository.git`. No more messing around with requirements.txt or manually resolving dependency conflicts.
 - build your own applications using a rich set of approaches, such as `EmbeddedComfyClient` to run comfyui prompts directly in your python application without a backend; `AsyncRemoteComfyClient` to connect to a remote server, or a fully-featured OpenAPI 3.1 spec to make a rich UI.

#### Previous Discussion

This includes the setup.py changes: (resolves https://github.com/comfyanonymous/ComfyUI/issues/1613 https://github.com/comfyanonymous/ComfyUI/issues/1342 https://github.com/comfyanonymous/ComfyUI/issues/188)

 - setup.py now works
 - always installs the correct gpu accelerated pytorch for your platform, resolving issues like https://github.com/comfyanonymous/ComfyUI/issues/1759 https://github.com/comfyanonymous/ComfyUI/issues/1753 https://github.com/comfyanonymous/ComfyUI/issues/1737 https://github.com/comfyanonymous/ComfyUI/issues/1694 https://github.com/comfyanonymous/ComfyUI/issues/1164 https://github.com/comfyanonymous/ComfyUI/issues/1066 https://github.com/comfyanonymous/ComfyUI/issues/1130 https://github.com/comfyanonymous/ComfyUI/issues/813
 - provides a way for other packages to export custom nodes. they should have entry points of the form

**setup.py**
```python
setup(...,
entry_points={
        # key is required
        'comfyui.custom_nodes': [
            'mypackagename = module.root.to.recursively.scan.for.nodes',
        ],
    })
```

all .py files with
```
NODE_CLASS_MAPPINGS = {
    "BinaryPreprocessor": Binary_Preprocessor
}
NODE_DISPLAY_NAME_MAPPINGS = {
    "BinaryPreprocessor": "Binary Lines"
}
```
that are in a package (i.e. a directory with an `__init__.py`) in `module.root.to.recursively.scan.for.nodes` or deeper will be recognized by comfyui as providing nodes.

since you will be using pip to install custom nodes this way, you will get dependency resolution automatically, so you never have to `pip install -r requirements.txt` or similar garbage. this will deprecate / obsolete ComfyUI-Manager's functionality, you can use any pip package manager UI (mostly resolves https://github.com/comfyanonymous/ComfyUI/issues/1773). this also fixes issues like https://github.com/comfyanonymous/ComfyUI/issues/1678 https://github.com/comfyanonymous/ComfyUI/issues/1665 https://github.com/comfyanonymous/ComfyUI/issues/1596 https://github.com/comfyanonymous/ComfyUI/issues/1385 https://github.com/comfyanonymous/ComfyUI/issues/1373 if adopted

you would also be able to add the comfyui git hash and custom nodes packages by git+commit or name in the metadata for maximum reproducibility.

This PR also creates two new api endpoints documented in the openapi.yaml. These are ergonomic for common usage like batches or using from other applications.

test using the instructions from: https://github.com/hiddenswitch/ComfyUI/tree/master#manual-install-windows-linux-macos

also copied here to improve comprehension of the purpose of this PR

## Manual Install (Windows, Linux, macOS)

You must have Python 3.11 or 3.10 installed. 3.12 is not yet supported. On Windows, download the latest Python from their website. You can also [directly download 3.11.4 here.](https://www.python.org/ftp/python/3.11.4/python-3.11.4-amd64.exe).

On macOS, install exactly Python 3.11 using `brew`, which you can download from https://brew.sh, using this command: `brew install python@3.11`. Do not use 3.9 or older, and do not use 3.12 or newer. Its compatibility with Stable Diffusion in both directions is broken. 

1. Create a virtual environment:
      ```shell
      python -m virtualenv venv
      ```
2. Activate it on
   **Windows (PowerShell):**
   ```pwsh
   Set-ExecutionPolicy Unrestricted -Scope Process
   & .\venv\Scripts\activate.ps1
   ```
   
   **Linux and macOS**
   ```shell
   source ./venv/bin/activate
   ```

3. Then, run the following command to install `comfyui` into your current environment. This will correctly select the version of pytorch that matches the GPU on your machine (NVIDIA or CPU on Windows, NVIDIA AMD or CPU on Linux):
    ```shell
    pip install git+https://github.com/hiddenswitch/ComfyUI.git
    ```
4. To run the web server:
    ```shell
    comfyui
    ```
    To generate python OpenAPI models:
    ```shell
    comfyui-openapi-gen
    ```

## Manual Install (Windows, Linux, macOS) For Development

 1. Clone this repo:
    ```shell
    git clone https://github.com/comfyanonymous/ComfyUI.git
    cd ComfyUI
    ```

 2. Put your Stable Diffusion checkpoints (the huge ckpt/safetensors files) into the `models/checkpoints` folder. You can download SD v1.5 using the following command:
    ```shell
    curl -L https://huggingface.co/runwayml/stable-diffusion-v1-5/resolve/main/v1-5-pruned-emaonly.ckpt -o ./models/checkpoints/v1-5-pruned-emaonly.ckpt
    ```

 3. Put your VAE into the `models/vae` folder.

 4. (Optional) Create a virtual environment:
    1. Create an environment:
       ```shell
       python -m virtualenv venv
       ```
    2. Activate it:
       
       **Windows (PowerShell):**
       ```pwsh
       Set-ExecutionPolicy Unrestricted -Scope Process
       & .\venv\Scripts\activate.ps1
       ```
       
       **Linux and macOS**
       ```shell
       source ./venv/bin/activate
       ```

 5. Then, run the following command to install `comfyui` into your current environment. This will correctly select the version of pytorch that matches the GPU on your machine (NVIDIA or CPU on Windows, NVIDIA AMD or CPU on Linux):
    ```shell
    pip install -e .
    ```
 6. To run the web server:
    ```shell
    comfyui
    ```
    To generate python OpenAPI models:
    ```shell
    comfyui-openapi-gen
    ```
    
    You can use `comfyui` as an API. Visit the [OpenAPI specification](comfy/api/openapi.yaml). This file can be used to generate typed clients for your preferred language.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,11 +1,27 @@
...
-accelerate
-pyyaml
+pytorch-lightning>=2.0.0
+aiohttp>=3.8.4
+accelerate>=0.18.0
+pyyaml>=6.0
+scikit-image>=0.20.0
+jsonmerge>=1.9.0
+clean-fid>=0.1.35
+clip>=0.2.0
```
Comment (tildebyte) [inline requirements.txt]: Are you *certain* that you mean https://pypi.org/project/clip v0.2.0?

**Comment by tildebyte** [comment]
This looks really good (worked well for me).

**Comment by doctorpangloss** [comment]
This is now tested on fairly complex setups **except** AMD DirectML on Windows. I will try that on my AMD Windows station later. Additionally, PyTorch 2.1+ nightlies are now the default for all accelerators I tested (Windows NVIDIA, Windows CPU, Linux NVIDIA, Linux AMD, macOS Apple Silicon). This leaves Windows AMD and macOS AMD to test.

**Comment by tildebyte** [comment]
> PyTorch 2.1+ nightlies

Looks like for NVidia, I currently get torch@https://download.pytorch.org/whl/cu118/torch-2.0.1%2Bcu118-cp311-cp311-win_amd64.whl (the url in setup.py is for [torch 2.0.1 and CUDA 11.8](https://github.com/doctorpangloss/ComfyUI/blob/9d18b0852f36b633fa06fa0c565b821a3bbbceb2/setup.py#L35C23-L35C61)) when I run `pip install -e .`, not 2.1+nightly (unless maybe there's a dependency which I'm not seeing which causes torch==2.1)

**Comment by doctorpangloss** [comment]
That's the error! I accidentally left out that commit, I'll tie up my other work shortly and tidy this up.

**Comment by mcmonkey4eva** [comment]
This PR makes a large amount of unrelated changes that makes it extremely difficult to review or accept, eg reformatting of unrelated code.
Could you please resubmit only the core/relevant changes needed for your goal here?

**Comment by tildebyte** [comment]
Just to throw this out there - I've been running Comfy with this setup and (important part) Python 3.11 for weeks now without issue.

Unrelated, but, if I were a reviewer here, I'd at least ask that the API stuff be split into a separate PR - it's cool and all, but not really related to the bulk of what's here.

**Comment by doctorpangloss** [comment]
> I'd at least ask that the API stuff be split into a separate PR

that can be done for sure. imo it's nice to deliver immediate ROI when you take your long term ROI packaging bitter medicine

File: README.md [hunk 2]
```diff
@@ -105,7 +105,40 @@ See the [Config file](extra_model_paths.yaml.example) to set the search paths fo
...
+   ```
+   
+   **Linux and macOS**
+   ```shell
+   source ./venv/bin/activate
+   ```
+
+3. Then, run the following command to install `comfyui` into your current environment. This will correctly select the version of pytorch that matches the GPU on your machine (NVIDIA or CPU on Windows, NVIDIA AMD or CPU on Linux):
+    ```shell
+    pip install git+https://github.com/hiddenswitch/ComfyUI.git
```
Comment (tildebyte) [inline README.md:91]: Don't forget to point this to the main repo (i.e. *this* repo)

Comment (NeedsMoar) [reply README.md:91]: Also don't forget there are weirdos like me who have a RTX4090 and a 7900XTX in the same machine.   I expect AMD will have MIOpen ported so the torch people can get a ROCm version out around the time I'm dead of old age but I definitely wouldn't want to suddenly get switched back to directml.   :-)

**Comment by oxc** [comment]
It would be nice to be able to select the target system manually, not only through auto detection. One might want to provide a list of prebuilt packages, on a build machine that has none of the GPUs it is building for.

**Comment by doctorpangloss** [comment]
> One might want to provide a list of prebuilt packages

that's what `pip`'s dependency management is. In any case, you can always install your preferred version of torch, and everything will just work.

**Comment by Chaoses-Ib** [comment]
The readme about authoring custom nodes uses `setup.py` to store project metadata. How about storing the metadata in `pyproject.toml`? It's more general, one can choose other backends like [Hatching](https://github.com/pypa/hatch), [PDM](https://github.com/pdm-project/pdm) and [Flit](https://github.com/pypa/flit) by simply changing `build-system`.

The entry points can be specified like
```toml
[project.entry-points."comfyui.custom_nodes"]
mypackage = "mypackage_custom_nodes"
```

File: comfy/nodes/package.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                                ]),
+                                ExportedNodes())
+        custom_nodes_mappings = ExportedNodes()
+        if custom_nodes is not None:
+            custom_nodes_mappings.update(_import_and_enumerate_nodes_in_module(custom_nodes, print_import_times=True))
+
+        # load from entrypoints
+        for entry_point in iter_entry_points(group='comfyui.custom_nodes'):
+            # Load the module associated with the current entry point
+            module = entry_point.load()
```
Comment (Chaoses-Ib) [inline comfy/nodes/package.py:93]: Modules may raise exceptions during load. Catching exceptions here can prevent breaking other imports.

**Pull request closed without merging**